plugin/pkg/auth/authenticator/password/keystone/keystone.go

```diff
@@ -45,18 +50,74 @@ func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticatePassword(usernam
 		glog.Info("Failed: Starting openstack authenticate client")
 		return nil, false, errors.New("Failed to authenticate")
 	}
-
+//FIXME user_id, not user_name...
 	return &user.DefaultInfo{Name: username}, true, nil
 }
 
-// NewKeystoneAuthenticator returns a password authenticator that validates credentials using openstack keystone
-func NewKeystoneAuthenticator(authURL string) (*KeystoneAuthenticator, error) {
-	if !strings.HasPrefix(authURL, "https") {
-		return nil, errors.New("Auth URL should be secure and start with https")
+func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticateToken(value string) (user.Info, bool, error) {
+//FIXME kfox1111: To enhance performance, cache admin token/client.
+	cfg, provider, err := loadConfig(keystoneAuthenticator.configPath)
+	if err != nil {
+		return nil, false, errors.New("Internal error getting the Keystone provider")
+	}
+
+	client, err := openstack.NewIdentityAdminV3(provider, gophercloud.EndpointOpts{
+		Region: cfg.Global.Region,
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	response, err := client.Request("GET", client.ServiceURL("auth", "tokens"), gophercloud.RequestOpts{
+	        MoreHeaders: map[string]string{"X-Subject-Token": value},
+	        OkCodes:     []int{200},
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	defer response.Body.Close()
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		glog.V(3).Infof("Cannot get HTTP response body from keystone token validate: %v", err)
+		return nil, false, err
+	}
+	obj := struct{ Token struct { User struct { Id string }}}{}
+	err = json.Unmarshal(bodyBytes, &obj)
+	if err != nil {
+		return nil, false, err
+	}
+	return &user.DefaultInfo{Name: obj.Token.User.Id}, response.StatusCode == 200, nil
```

Comment:
> it may be better to just do the little bit of authz that may be needed right in the code here...

Depending on what you mean by that, I don't think we want to mix the k8s authn/authz architecture that way here. The information about the namespace and what operation is being performed isn't available to the auth layer

I'm not very familiar with keystone, but it seems like there is not "one true way" to determine allowed kubernetes permissions from keystone (keystone projects might map 1:1 to k8s namespaces or might not, keystone roles don't have inherent meaning within kubernetes, etc). 

> Users would then expect that they could login and do stuff with their cluster, but would not want anyone not in their tenant to be able to login to their Kubernetes clulster.

Limiting to a particular tenant would also imply limiting to a particular domain, which would mean user names would actually be unique and remove the need for domain qualification 

> Checks that the namespace specified = the tenant id out of the keystone token and the role is associated. if no namespace is specified, use the namespace = the tenant id instead of 'default'.

we should not allow "no namespace" to possibly mean anything other than default (and I'm pretty sure it only possibly means that during create)… that becomes really really hard to reason about.